### PR TITLE
Running remote queries: Save repo lists in VS Code config

### PR DESCRIPTION
**Pull request by shati-patel**
This PR tweaks how you run remote queries 😎 
Previously, each query needed a `.repositories` file that listed which repositories you wanted to analyze.

Now, you can save pre-defined "repository lists" in your VS Code settings, and choose one of those when you run a remote query. 

New flow:
1. If you still have a `.repositories` file, we use the information from there to run your remote query.
2. Otherwise, if you have saved repository lists in your VS Code settings, you can select one of those to run against.
3. If not, you can enter a single repo in a text input box and we run against that.

The `getRepositories` function should give an error if you end up with an empty string/array/object. It doesn't check if your repositories are valid, but the API call itself gives an error in that case (e.g. ❌ `"HttpError: Repository 'xxx' not found"`).

Question:
- Since this isn't a secret, is it okay for this internal setting to be visible in the Settings UI? (@jf205?)
image
  I could hide it completely by omitting the [package.json](https://github.com/github/vscode-codeql/pull/918/files#diff-be4c5a31505df7de81fa4900e84e66c04d1a6bca9b3da50f8061d2cb2bfbbde1) change, but that would remove the validation (that the setting has the correct "shape", i.e. an object where the values are arrays of strings)

## Checklist

N/A: this is part of an internal-only feature (see linked issue for more context/details).

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] `@github/docs-content-codeql` has been cc'd in all issues for UI or other user-facing changes made by this pull request.

### CODE REVIEW

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-remote-query.ts:16]: It looks like this gets the repository lists once and then it is fixed. What happens if someone changes their repository lists? Do they need to restart?

Maybe, instead of a list, call `getRemoteRepositoryLists()` every time immediately before you use it.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-remote-query.ts:61]: Perhaps better to also check that `REPO_LISTS` is not an empty object, eg-

```suggestion
  if (Object.keys(REPO_LISTS || {}).length) {
```

Also, see my comment above. This might be better:

```suggestion
  if (Object.keys(getRemoteRepositoryLists() || {}).length) {
```

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-remote-query.ts:67]: nit: Maybe a little cleaner.

```suggestion
    const quickPickItems = Object.enties(REPO_LISTS).map<RepoListQuickPickItem>(([key, value]) => (
      {
        label: key,                // the name of the repository list
        repoList: value, // the actual array of repositories
      }
    ));
```

Comment (aeisenberg) [reply extensions/ql-vscode/src/run-remote-query.ts:67]: Also, note that repositories lists is a user setting and it may be malformed. Just make sure that if a malformed list (eg- a single string or even a number) is passed, then the error is expressed to the user in an understandable way.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-remote-query.ts:74]: nit:

```suggestion
    if (quickpick?.repoList.length > 0) {
```

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-remote-query.ts:110]: Not something to handle now, but this means that users can only choose a ref if they have a .repositories file. If there is no file, then they can only choose `main`. This is fine for now, but we should consider making this more flexible this later.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-remote-query.ts:116]: Since `findLanguage` can take a while to run and `getRepositories` is blocked by a user event, here's a trick to get them to run in "parallel":

```suggestion
    [ language, repositories ] = await Promise.all([findLanguage(cliServer, uri), getRepositories()]);;
```

`Promise.all` accepts an array of promises and resolves when all of its arguments resolve. Or, it rejects if any one of them reject. The return value is an array containing all of the resolved values of each promise.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (shati-patel) [reply extensions/ql-vscode/src/run-remote-query.ts:72]: I tried this and got an error in the lines below that "Object `quickpick` might be undefined" 😅

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (shati-patel) [reply extensions/ql-vscode/src/run-remote-query.ts]: > Also, note that repositories lists is a user setting and it may be malformed. Just make sure that if a malformed list (eg- a single string or even a number) is passed, then the error is expressed to the user in an understandable way.

👍🏽 Good point. I've re-instated the [package.json](https://github.com/github/vscode-codeql/pull/918/files#diff-be4c5a31505df7de81fa4900e84e66c04d1a6bca9b3da50f8061d2cb2bfbbde1) change, which at least validates that each list is an array of strings. (I removed it from the package.json initially since that stops it from showing up in the settings UI, but maybe we don't need to be too secretive 🤫 )

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (shati-patel) [reply extensions/ql-vscode/src/run-remote-query.ts:108]: 👍🏽 I expect that we won't need the `ref` at all eventually, though I will double check with the others 😃 
It just specifies from which branch of the controller repository we're running the "remote query" workflow. I can't see why it should ever be different from `main` (except for testing).

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [reply extensions/ql-vscode/src/run-remote-query.ts:72]: Hmmm...ok. Not sure what's happening, but fine to ignore.

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (aeisenberg) [inline extensions/ql-vscode/src/run-remote-query.ts:58]: nit: Typically, if something is not a module-level constant, then we avoid using CAPs for it.

```suggestion
  const repoLists = getRemoteRepositoryLists();
```

**Review by aeisenberg — COMMENTED** [review]
Looking good!

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (shati-patel) [reply extensions/ql-vscode/src/run-remote-query.ts:108]: (Confirmed with @robertbrignull that we don't need to worry about easily setting the `ref` in VS Code. If required for testing, we can send a manual API request.)

File: extensions/ql-vscode/src/config.ts
Comment (edoardopirovano) [inline extensions/ql-vscode/src/config.ts:302]: Extreme nit:
```suggestion

```

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (edoardopirovano) [inline extensions/ql-vscode/src/run-remote-query.ts:87]: Should we also do some basic validation here to ensure the format is indeed `<owner>/<repo>` and fail early (and with a useful error message) if this is indeed the case?

**Review by edoardopirovano — COMMENTED** [review]
Some very minor comments but otherwise looks good!

File: extensions/ql-vscode/src/run-remote-query.ts
Comment (shati-patel) [reply extensions/ql-vscode/src/run-remote-query.ts:88]: Thanks for the help on Slack ✨ Cross-posting the regex you shared here for reference: https://regex101.com/r/krDHm5/2

Comment (edoardopirovano) [reply extensions/ql-vscode/src/run-remote-query.ts:88]: No problem! For future reference, I would also add a comment documenting what the regex is matching, i.e. strings of the form `owner/repo` where:
- `owner` is made up of alphanumeric characters or _single_ hyphens, starting and ending in an alphanumeric character
- `repo` is made up of alphanumeric characters, hyphens, or underscores

**Comment by edoardopirovano** [comment]
I'm happy with this once the above comment has been added, will leave it to @aeisenberg to approve if he's happy his comments have been addressed.

**Comment by shati-patel** [comment]
Oops, I've introduced a high-severity code scanning alert 😅 I'll try to figure out what I need to fix first... 👀

**Comment by edoardopirovano** [comment]
> Oops, I've introduced a high-severity code scanning alert sweat_smile I'll try to figure out what I need to fix first... eyes

Ah, yes, Code Scanning is not too happy with our regex... I think changing it to `^(?:[a-zA-Z0-9]+-)*[a-zA-Z0-9]+\/[a-zA-Z0-9-_]+$` should get rid of the alert and matches the same things (see https://regex101.com/r/35eyb8/1).

**Review by edoardopirovano — APPROVED** [review]

**Comment by jf205** [comment]
>Since this isn't a secret, is it okay for this internal setting to be visible in the Settings UI? 

👍🏻 (thanks for checking!)

**Review by aeisenberg — APPROVED** [review]
Apologies for taking so long to review. This looks good.

**Pull request merged**
